### PR TITLE
Segmentation Pretrained Weights

### PR DESCRIPTION
**Pull request by isaaccorley**
This PR addresses part of #1044 adds the ability to load pretrained weights from a backbone model e.g. ResNet into a semantic segmentation encoder. This works for the segmentation-models-pytorch Unet and DeepLabv3 implementations but not the FCN because we aren't using a backbone for that.

### CODE REVIEW

**Comment by adamjstewart** [comment]
This only supports ImageNet pretrained weights, which is pretty useless for us. We really want support for our weight enums like we have for regression/classification/byol. Could we manually load weights from state dict ourselves like we do with our timm-based models?

**Comment by adamjstewart** [comment]
Oh wait, I didn't read the full PR. Maybe this does cover all the features we want. You'll definitely want to update the docstring though.

**Comment by adamjstewart** [comment]
Also needs tests

**Comment by nilsleh** [comment]
Maybe also put a note in the docs that not all our pretrained models (i.e. the ViTs) will be compatible according to the [smp docs](https://github.com/qubvel/segmentation_models.pytorch#timm-encoders-) or maybe there is a way around, not sure.

**Comment by isaaccorley** [comment]
The other thing I realized is that by default smp will use torchvision for resnet backbones e.g. `resnet18`. If you want to force using timm implementations you have to prefix the encoder names with `tu-` e.g. `tu-resnet18`.

**Comment by calebrob6** [comment]
We could just only allow timm pretrained backbones here

**Comment by isaaccorley** [comment]
This actually needs #1049 to be solved otherwise this only supports resnet backbones.

**Comment by isaaccorley** [comment]
> We could just only allow timm pretrained backbones here

So do we want to prefix `tu-` to all the backbones or do we want the user to do it?

**Comment by adamjstewart** [comment]
That is a tough question. I would be fine with only supporting timm backbones for now, which would suggest prepending `tu-` for the user so that it matches other trainers. But there may come a time in which we want to support non-timm backbones because pretrained models are not always made with timm.

**Review by calebrob6 — APPROVED** [review]

**Comment by calebrob6** [comment]
We could prefix "tu-" if it results in a valid timm backbone to give ourselves room to grow.

File: torchgeo/trainers/segmentation.py [hunk 1]
```diff
@@ -80,13 +85,23 @@ def config_task(self) -> None:
...
+                self.model.encoder = utils.load_state_dict(self.model, state_dict)
+
     def __init__(self, **kwargs: Any) -> None:
         """Initialize the LightningModule with a model and loss function.
 
         Keyword Args:
             model: Name of the segmentation model type to use
             backbone: Name of the timm backbone to use
-            weights: None or "imagenet" to use imagenet pretrained weights in
+            weights: None or True to use imagenet pretrained weights in
```
Comment (calebrob6) [inline torchgeo/trainers/segmentation.py:104]: Probably need to say this can be a string and stuff

**Comment by calebrob6** [comment]
Ugh, also the horrific model checkpoint tests

**Comment by calebrob6** [comment]
We need this :)

**Comment by calebrob6** [comment]
@isaaccorley does this work with our custom ResNet weights?

**Comment by isaaccorley** [comment]
I haven't tried using the weights enums but in theory it should since the encoders are just resnet models.

**Comment by adamjstewart** [comment]
It should be the same loading code and docstring description as all of the other trainers.

**Comment by isaaccorley** [comment]
I just made it work with the ResNet weights

File: tests/trainers/test_segmentation.py [hunk 2]
```diff
@@ -111,6 +122,50 @@ def model_kwargs(self) -> dict[Any, Any]:
             "ignore_index": 0,
         }
 
+    @pytest.fixture(
+        params=[
+            weights
+            for model in list_models()
+            for weights in get_model_weights(model)
+            if "resnet" in weights.meta["model"]
```
Comment (isaaccorley) [inline tests/trainers/test_segmentation.py:130]: SMP doesn't support ViT weights for the encoder yet (only MiT)

File: torchgeo/trainers/segmentation.py [hunk 3]
```diff
@@ -100,8 +119,9 @@ def __init__(self, **kwargs: Any) -> None:
         Keyword Args:
             model: Name of the segmentation model type to use
             backbone: Name of the timm backbone to use
-            weights: None or "imagenet" to use imagenet pretrained weights in
-                the backbone
+            weights: Either a weight enum, the string representation of a weight enum,
+                True for ImageNet weights, False or None for random weights,
+                or the path to a saved model state dict.
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py:124]: Need versionchanged and need to document that FCN does not support pretrained models.

**Review by calebrob6 — DISMISSED** [review]

File: tests/conf/inria.yaml [hunk 4]
```diff
@@ -3,7 +3,7 @@ module:
   loss: "ce"
   model: "unet"
   backbone: "resnet18"
-  weights: "imagenet"
+  weights: true
```
Comment (adamjstewart) [inline tests/conf/inria.yaml:6]: Does this download from the internet? We want to avoid that

File: tests/trainers/test_segmentation.py [hunk 5]
```diff
@@ -111,6 +122,53 @@ def model_kwargs(self) -> dict[Any, Any]:
...
+        model_kwargs["weights"] = mocked_weights
+        SemanticSegmentationTask(**model_kwargs)
+
+    def test_weight_str(
+        self, model_kwargs: dict[str, Any], mocked_weights: WeightsEnum
+    ) -> None:
+        model_kwargs["backbone"] = mocked_weights.meta["model"]
+        model_kwargs["in_channels"] = mocked_weights.meta["in_chans"]
+        model_kwargs["weights"] = str(mocked_weights)
+        SemanticSegmentationTask(**model_kwargs)
```
Comment (adamjstewart) [inline tests/trainers/test_segmentation.py:170]: Can you add tests that test the real weights? See all other trainer tests for examples. It's the things marked as slow that only run during the release and are allowed to access the internet to test real world integration.

File: torchgeo/trainers/segmentation.py [hunk 6]
```diff
@@ -127,6 +144,9 @@ class and used with 'ce' loss
             The *class_weights*, *freeze_backbone*,
             and *freeze_decoder* parameters.
 
+        .. versionchanged:: 0.5
+           The *weights* parameter supports WeightEnums and checkpoint paths.
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py:148]:
```suggestion
           The *weights* parameter now supports WeightEnums and checkpoint paths.
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**